### ISSUE TITLE
Linear data support, gpu acceleration

### ISSUE THREAD

**Issue by naixx** [charvey2718/nox#2]
I've tried the tool and it works much better then the two competitors in some scenarios. I'd say almost perfect (just a few artifacts that starnet usually produces)
I also want to mention that I managed to run the app on GPU. After playing with parameters, the app runs extremely fast compared to Starnet or Starx. It is worth mention in readme that a user can replace tensorflow lib and provide cuda stuff to launch removal on gpu.  I've also used batch parameter of 8 to speed up

What would be really useful is linear data support. In old days, there were scripts that applied mft and reverse mft while using starnet. I believe a trained will produce much better results. Do you plan to continue development of the nox? If no, what would be your suggestion? Unfortunately originals papers are not available, if there were details about linear data

**Comment by charvey2718** [comment]
Hello @naixx 
Thanks for raising these items. Great to hear you got some good results.

I have been quite busy of late but should now have a little more time again for personal projects like this.

I think I can add linear data support without too much fuss. I already have methods for this - it's just integration into the app and testing. The method would essentially be to auto-stretch linear data (by setting a flag), apply the star removal, and then reverse the stretch.

I also plan to improve the synthetic training data to include examples "star halos" which might help improve removal of some artefacts. I was also going to have a look at using an OTS in-painting model with some transfer learning for astro images. It could potentially give better results than the StarNet architecture.

Could you share some more details please on running on GPU and providing CUDA, please? Was this on Windows with a custom
TF build? I was unable to achieve this on Windows since the official TF builds with GPU support are only available for Linux. I'd be delighted to get speedups like you describe. I'll update the README as suggested!

**Comment by naixx** [comment]
Sure! Basically I've used tensorflow from pixinsight `bin/` as well as other cuda stuff from it. I believe you can get them from either starnet, starxterminator or other sources. But it was not enought, additional cuda libs were required which i downloaded from https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/windows-x86_64/cudnn-windows-x86_64-8.9.4.25_cuda11-archive.zip

```
cublas64_11.dll
cublasLt64_11.dll
cudart64_110.dll
cudnn_adv_infer64_8.dll
cudnn_adv_train64_8.dll
cudnn_cnn_infer64_8.dll
cudnn_cnn_train64_8.dll
cudnn_ops_infer64_8.dll
cudnn_ops_train64_8.dll
cudnn64_8.dll
cufft64_10.dll
cusparse64_11.dll
libwinpthread-1.dll
tensorflow.dll
zlibwapi.dll
```
After that, processing with batch parameter went extremely fast. Almost one or two seconds(but loading file into gpu(?)/initialization took additional 5-7). While starnet or starx do it in 15-20 seconds overall I find it extremely fast.

A few issues I observed: if patch and stride are equal, one can observe a visible patch truncation. Another issue is that very edges of the images are modified despite star absence. Combining those two, you can get a case where path in the middle can produce articafts. When stride and patch are different, image is almost unnoticeable.
See the issue in the middle and right bottom artifacts
<img width="2339" height="1900" alt="Image" src="https://github.com/user-attachments/assets/0932823e-4a40-4e7f-8abc-4c8c2b67cb1d" />

And the third issue is that small galaxies are still eliminated in another example no matter I stretch image differently or optimize parameters.

So in noisy CFA case it works extremely well, I managed to produce almost no patch edges unlike starx, much less or almost none artifacts then starnet2 produces. With mono, I wasn't that successful

<img width="1920" height="1111" alt="Image" src="https://github.com/user-attachments/assets/c2dcf75b-e8d5-48e5-aa61-e3c14f782d3b" />

**Comment by charvey2718** [comment]
Thank you for those details! Greatly appreciated! I'll add them to the README.

In general, the stride should be less than the patch/tile size. The net can only work well away from patch edges (which is a side-effect of the convolution process). I don't think much can be done about that except to keep the stride less than the patch size to get overlapping tiles. Then the processed image is composed only of the central area from each tile, excluding areas near edges.

For this same reason, it may be necessary to crop the reassembled image slightly if there are edge artefacts around the perimeter. Each tile uses mirroring around its edges to perform the convolution, which helps, but edge artefacts still occur.

For galaxies, essentially the model can't tell the difference between them and stars. I can try to incorporate galaxy-like objects in the training data so the model learns not to remove these. I can't guess how successful this might be. I usually just mask galaxies to protect them.

Also on the to-do list is to increase the "cost" associated with doughnut-like artefacts in the discriminator, which will help produce better star removal results, and might help with galaxies too.

I'll try to add linear data capability within the next few days. (You raised this issue just as I was getting back to this project!)

**Comment by naixx** [comment]
Thank you for clarification. Yep, I also mask galaxies to avoid losses. 

I'm not an ai/ml engineer, I thought to do something similar to implement that you have done already(before I found the project). My idea was that after neural network processes the patch, we can calculate local psf of the patch. If "star" to delete doesn't deviate a lot to the psf, they we do actual deletion. This may sound newbie but do you think this is achievable on post nn step? For model that might be difficult to distinguish a star and a small galaxy

**Comment by charvey2718** [comment]
I've just released [v1.1.0](https://github.com/charvey2718/nox/releases/tag/v1.1.0) which offers temporary automatic stretching of linear data. There are now details in the [README](https://github.com/charvey2718/nox/blob/main/README.md) file.

I haven't had a chance to do much testing - I am without a sufficiently powerful personal PC at the moment. My preliminary checks, however, look promising. Please let me know how you get on.

I have also updated the README with your guidance on how to get GPU acceleration (thank you for that).

The DNN models are unchanged. That is, I have performed no more training for removing stars vs galaxies, or for star halo removal. I will need something more powerful than my old 2014-era i3 laptop (with no GPU and only 4GB of RAM)! I originally did this work on a Jetson Orin AGX.

Your idea to check removed stars against a PSF is possible I think. And then removed stars that significantly deviate from all forms of the PSF could be masked. However, I think removing halos (which deviate from the PSF too) would still present a challenge. I will keep thinking and try some things...

**Comment by naixx** [comment]
@charvey2718 thank you! it works! I wonder why filesize changes from 145Mb -> 89Mb

**Comment by charvey2718** [comment]
Ahh, very interesting — I didn't notice that.

Besides the new functionality, the main changes are updating to newer libraries, in particular OpenCV 4.12 (previously ~4.8) and a newer GCC compiler (15.2 vs ~13.1). The newer compiler and linker are probably better at avoiding unused code from static libraries, so much less OpenCV and C++ runtime code ends up in the final executable. Since OpenCV is very large and this project uses only a small subset of it, that likely explains the significantly smaller binary.

**Comment by naixx** [comment]
I should clarify that I was talking about original/processed image. Seems like processed one is much less than original. Is there some precision transformation?

**Comment by charvey2718** [comment]
Oh I see - yes, I completely misunderstood that.

The output file from nox should be 16-bit TIFF. (Nox's internal processing uses double precision, i.e., 64-bit.) If the input file has a higher bit depth (e.g, 32 bits), then the output file size will be smaller after processing. I will add a 32-bit output flag to nox as it's more relevant now than a couple of years ago - it's a couple of lines of code at most.

If your input file came from
PixInsight, then it could be a 32-bit TIFF.

**Comment by charvey2718** [comment]
I added the -B flag (in the current release) which accepts 16-, 32- or 64-bit output. The default is 16 bits. All internal processing is in 32 bits.

For non-linear data, 16 bits should be more than adequate. For linear data, there is an argument for 32 bits.

After running a few examples, some output files are bigger, some are smaller. I think it's because removing stars lowers the dynamic range and variance of pixel values. This can result in slightly more efficient TIFF storage. All image data is preserved.
